<a href="https://colab.research.google.com/github/tzmjoseph/UT_Austin_PGP_AIML_Projects/blob/main/ThomasJoseph_Full_Code_NLP_RAG_Project_Notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Problem Statement

### Business Context

The healthcare industry is rapidly evolving, with professionals facing increasing challenges in managing vast volumes of medical data while delivering accurate and timely diagnoses. The need for quick access to comprehensive, reliable, and up-to-date medical knowledge is critical for improving patient outcomes and ensuring informed decision-making in a fast-paced environment.

Healthcare professionals often encounter information overload, struggling to sift through extensive research and data to create accurate diagnoses and treatment plans. This challenge is amplified by the need for efficiency, particularly in emergencies, where time-sensitive decisions are vital. Furthermore, access to trusted, current medical information from renowned manuals and research papers is essential for maintaining high standards of care.

To address these challenges, healthcare centers can focus on integrating systems that streamline access to medical knowledge, provide tools to support quick decision-making, and enhance efficiency. Leveraging centralized knowledge platforms and ensuring healthcare providers have continuous access to reliable resources can significantly improve patient care and operational effectiveness.

**Common Questions to Answer**

**1. Diagnostic Assistance**: "What are the common symptoms and treatments for pulmonary embolism?"

**2. Drug Information**: "Can you provide the trade names of medications used for treating hypertension?"

**3. Treatment Plans**: "What are the first-line options and alternatives for managing rheumatoid arthritis?"

**4. Specialty Knowledge**: "What are the diagnostic steps for suspected endocrine disorders?"

**5. Critical Care Protocols**: "What is the protocol for managing sepsis in a critical care unit?"

### Objective

As an AI specialist, your task is to develop a RAG-based AI solution using renowned medical manuals to address healthcare challenges. The objective is to **understand** issues like information overload, **apply** AI techniques to streamline decision-making, **analyze** its impact on diagnostics and patient outcomes, **evaluate** its potential to standardize care practices, and **create** a functional prototype demonstrating its feasibility and effectiveness.

### Data Description

The **Merck Manuals** are medical references published by the American pharmaceutical company Merck & Co., that cover a wide range of medical topics, including disorders, tests, diagnoses, and drugs. The manuals have been published since 1899, when Merck & Co. was still a subsidiary of the German company Merck.

The manual is provided as a PDF with over 4,000 pages divided into 23 sections.

## Installing and Importing Necessary Libraries and Dependencies

In [ ]:
# Installation for GPU llama-cpp-python
# uncomment and run the following code in case GPU is being used
!CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python==0.2.28 --force-reinstall --no-cache-dir -q

# Installation for CPU llama-cpp-python
# uncomment and run the following code in case GPU is not being used
# !CMAKE_ARGS="-DLLAMA_CUBLAS=off" FORCE_CMAKE=1 pip install llama-cpp-python==0.2.28 --force-reinstall --no-cache-dir -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.4/9.4 MB 65.1 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 222.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 172.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.6/44.6 kB 193.7 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
numba 0.60.0 requires numpy<2.1,>=1.22, but you have numpy 2.4.4 which is incompatible.


**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

In [ ]:
# For installing the libraries & downloading models from HF Hub
!pip install huggingface_hub==0.35.3 pandas==2.2.2 tiktoken==0.12.0 pymupdf==1.26.5 langchain==0.3.27 langchain-community==0.3.31 chromadb==1.1.1 sentence-transformers==5.1.1 numpy==2.3.3 -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.1/62.1 kB 4.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.0/52.0 kB 5.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.0/44.0 kB 4.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 564.3/564.3 kB 22.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.1/24.1 MB 90.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 51.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 115.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.9/19.9 MB 100.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 486.6/486.6 kB 43.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 107.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.2/278.2 kB 25.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 99.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45

**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

In [ ]:
#Libraries for processing dataframes,text
import json,os
import tiktoken
import pandas as pd

#Libraries for Loading Data, Chunking, Embedding, and Vector Databases
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.document_loaders import PyMuPDFLoader
from langchain_community.embeddings.sentence_transformer import SentenceTransformerEmbeddings
from langchain_community.vectorstores import Chroma

#Libraries for downloading and loading the llm
from huggingface_hub import hf_hub_download
from llama_cpp import Llama

## Question Answering using LLM

#### Downloading and Loading the model

In [ ]:
# Loading the Mistral model
model_name_or_path = "TheBloke/Mistral-7B-Instruct-v0.2-GGUF"
model_basename = "mistral-7b-instruct-v0.2.Q6_K.gguf"

In [ ]:
model_path = hf_hub_download(
    repo_id=model_name_or_path,
    filename=model_basename
)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


mistral-7b-instruct-v0.2.Q6_K.gguf:   0%|          | 0.00/5.94G [00:00<?, ?B/s]

In [ ]:
llm = Llama(
    model_path=model_path,
    n_ctx=2300,
    n_gpu_layers=36,
    n_batch=512
)

AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 1 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | 


#### Response

In [ ]:
def response(query,max_tokens=512,temperature=0,top_p=0.95,top_k=50):
    model_output = llm(
      prompt=query,
      max_tokens=max_tokens,
      temperature=temperature,
      top_p=top_p,
      top_k=top_k
    )

    return model_output['choices'][0]['text']

In [ ]:
# Testing response with "What are the symptoms for type 2 diabetes?"
response("What are the symptoms for type 2 diabetes?")

"\n\nType 2 diabetes is a chronic condition that affects how your body metabolizes sugar (glucose) in your blood. Symptoms of type 2 diabetes can vary from person to person, and some people may not experience any noticeable symptoms at all until their diabetes has progressed significantly. However, common symptoms of type 2 diabetes include:\n\n1. Increased thirst and urination: Your body will try to get rid of the excess sugar in your blood by producing more urine, which can lead to increased thirst.\n2. Fatigue: High levels of glucose in your blood can make it difficult for your cells to get the energy they need, leading to feelings of weakness and fatigue.\n3. Blurred vision: High blood sugar levels can cause the lens in your eyes to swell, which can lead to blurry vision or even temporary blindness.\n4. Numbness or tingling in hands or feet: Damage to the nerves in your extremities can cause numbness, tingling, or pain in your hands and feet.\n5. Slow-healing wounds: High blood sug

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
# Defining user input for "What is the protocol for managing sepsis in a critical care unit?" and getting the response
user_input_1 = "What is the protocol for managing sepsis in a critical care unit?"
response(user_input_1)

Llama.generate: prefix-match hit


'\n\nSepsis is a life-threatening condition that can arise from an infection, and it requires prompt recognition and aggressive management in a critical care unit. The following are general steps for managing sepsis in a critical care unit:\n\n1. Early recognition: Recognize the signs and symptoms of sepsis early and initiate treatment as soon as possible. Sepsis can present with various clinical features, including fever or hypothermia, tachycardia or bradycardia, altered mental status, respiratory distress, and lactic acidosis.\n2. ABCs: Ensure airway patency, adequate breathing, and circulatory support. Provide high-flow oxygen via a non-rebreather mask or endotracheal tube if necessary. Initiate intravenous fluids to maintain adequate blood pressure and organ perfusion.\n3. Antibiotics: Administer broad-spectrum antibiotics as soon as possible based on the suspected source of infection and local microbiology data. Consider obtaining cultures before administering antibiotics if clin

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
# Defining the user input for "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what
# surgical procedure should be followed to treat it?" and getting the response
user_input_2 = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
response(user_input_2)

Llama.generate: prefix-match hit


"\n\nAppendicitis is a medical condition characterized by inflammation of the appendix, a small pouch-like structure that extends from the large intestine. The symptoms of appendicitis can vary from person to person, but some common signs include:\n\n1. Abdominal pain: The pain is typically located in the lower right side of the abdomen and may be constant or come and go. It may start as a mild discomfort that worsens over time.\n2. Loss of appetite: People with appendicitis often lose their appetite due to abdominal pain and nausea.\n3. Nausea and vomiting: Vomiting is a common symptom of appendicitis, especially in the early stages.\n4. Fever: A fever may be present, particularly if the appendix has ruptured or perforated.\n5. Constipation or diarrhea: Some people with appendicitis experience constipation, while others have diarrhea.\n6. Abdominal swelling: The abdomen may become swollen and tender to the touch.\n7. Inability to pass gas: Passing gas can be difficult due to the press

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
# Defining the user input for "What are the effective treatments or solutions for addressing sudden patchy hair loss,
# commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?" and getting the response
user_input_3 = (
 "What are the effective treatments or solutions for addressing sudden patchy hair loss,"
 "commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
)
response(user_input_3)

Llama.generate: prefix-match hit


"\n\nSudden patchy hair loss, also known as alopecia areata, is a common autoimmune disorder that affects the hair follicles. It can result in round or oval bald patches on the scalp, but it can also occur on other parts of the body such as the beard area, eyebrows, or eyelashes.\n\nThe exact cause of alopecia areata is not known, but it's believed to be related to a problem with the immune system. Some possible triggers for this condition include stress, genetics, viral infections, and certain medications.\n\nThere are several treatments that have been shown to be effective in addressing sudden patchy hair loss:\n\n1. Corticosteroids: These are anti-inflammatory drugs that can help reduce inflammation and suppress the immune system's attack on the hair follicles. They can be applied topically or taken orally, depending on the severity of the condition.\n2. Minoxidil: This is a medication that has been shown to promote hair growth in some people with alopecia areata. It works by increa

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
# Defining the user input for "What treatments are recommended for a person who has sustained a physical injury
# to brain tissue, resulting in temporary or permanent impairment of brain function?" and getting the response
user_input_4 = (
    "What treatments are recommended for a person who has sustained a physical injury"
    "to brain tissue, resulting in temporary or permanent impairment of brain function?"
)
response(user_input_4)

Llama.generate: prefix-match hit


'\n\nThere is no one-size-fits-all answer to this question as the specific treatment recommendations for a person with a brain injury depend on various factors such as the severity and location of the injury, the age and overall health of the individual, and the extent of the impairment. However, I can provide some general information about common treatments for brain injuries.\n\n1. Medical care: The first priority is to ensure that the person receives appropriate medical care to address any life-threatening conditions or complications related to the injury. This may include surgery to remove hematomas or other lesions, administration of medications to manage symptoms such as swelling or seizures, and close monitoring in an intensive care unit.\n2. Rehabilitation: Depending on the severity and location of the brain injury, rehabilitation may be necessary to help the person regain lost function and improve overall quality of life. Rehabilitation may include physical therapy to address 

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
# Defining the user input for "What are the necessary precautions and treatment steps for a person who has
# fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
# and getting the response
user_input_5 = (
    "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip,"
    "and what should be considered for their care and recovery?"
)
response(user_input_5)

Llama.generate: prefix-match hit


"\n\nA fractured leg is a serious injury that requires prompt medical attention. Here are some general guidelines for the necessary precautions and treatment steps for a person with a fractured leg during a hiking trip:\n\n1. Assess the situation: If someone has sustained a fracture while hiking, the first step is to assess the situation. Check if the person is in immediate danger or if there are any other injuries. Ensure that the area is safe for both you and the injured person before attempting to move them.\n2. Immobilize the leg: Use a splint or a makeshift sling to immobilize the leg as best as possible. This will help prevent further damage and reduce pain. Be careful not to apply too much pressure on the injury, as this can cause additional discomfort or even worsen the fracture.\n3. Provide first aid: Apply a clean dressing to the wound, if there is one, and apply gentle pressure to control any bleeding. If the person is experiencing significant pain or swelling, administer pa

**Observation**


*   Answers become more specific to the context provided.
*   We see better grounding, Higher faithfulness, Lower hallucination, and More source adherence when using LLM with Prompt Engineering.


## Question Answering using LLM with Prompt Engineering






In [ ]:
# Setting the system prompt
system_prompt = """
You are a highly reliable AI assistant to answer medical questions. Your task is to answer user questions.

Rules:
- Do not hallucinate or fabricate information.
- If there is no answer, say:
  "I coud not find any information regarding this."
- Prefer accuracy over completeness.
- Be concise but informative.
- Structure responses clearly.

Answering strategy:
1. Identify relevant parts of the context.
2. Synthesize the answer.
3. If needed, explain briefly.
"""

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
# Defining user input for "What is the protocol for managing sepsis in a critical care unit?" that is prefixed with
# system prompt and getting the response
user_input_1 = system_prompt + "\n" + "What is the protocol for managing sepsis in a critical care unit?"
response(user_input_1)

Llama.generate: prefix-match hit


'\n\nAnswer:\nThe protocol for managing sepsis in a critical care unit involves early recognition and prompt intervention. Here are the key steps:\n\n1. Recognition: Identify sepsis based on clinical signs, such as fever or hypothermia, tachycardia or bradycardia, respiratory distress, altered mental status, and lactic acidosis. Use the Sequential [Sepsis-related] Organ Failure Assessment (SOFA) score to assess organ dysfunction.\n\n2. Fluid resuscitation: Administer intravenous fluids to maintain adequate tissue perfusion. Aim for a mean arterial pressure (MAP) of 65 mmHg or higher and a central venous oxygen saturation (ScvO2) of >70%.\n\n3. Antibiotics: Start broad-spectrum antibiotics as soon as possible based on suspected infection source and local microbial resistance patterns.\n\n4. Source control: Address the underlying infection source, such as drainage of abscesses or debridement of necrotic tissue.\n\n5. Vasopressors: Use vasopressors if needed to maintain adequate blood pre

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
# Defining the user input for "What are the common symptoms for appendicitis, and can it be cured via medicine? If not,
# what surgical procedure should be followed to treat it?" that is prefixed with system prompt and getting the response
user_input_2 = system_prompt + "\n" + (
    "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure"
    "should be followed to treat it?"
)
response(user_input_2)

Llama.generate: prefix-match hit


"\n\nThe common symptoms for appendicitis include:\n1. Sudden pain in the lower right abdomen, often starting around the navel and then moving to the right side.\n2. Loss of appetite.\n3. Nausea and vomiting.\n4. Fever (often over 101°F or 38.3°C).\n5. Abdominal swelling and tenderness.\n6. Diarrhea or constipation.\n7. Inability to pass gas or have a bowel movement.\n\nAppendicitis cannot be cured via medicine alone, as the appendix will eventually rupture if left untreated. Rupture can lead to peritonitis, a serious inflammation of the abdominal cavity that requires immediate medical attention. The standard treatment for appendicitis is an appendectomy, which involves surgically removing the appendix. This procedure can be performed as an open surgery or laparoscopically (keyhole surgery). Both methods have their advantages and disadvantages, and the choice depends on various factors such as the patient's overall health, the severity of the condition, and the surgeon's expertise.\n\n

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
# Defining the user input for "What are the effective treatments or solutions for addressing sudden patchy hair loss,
# commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?" and getting the response
user_input_3 = system_prompt + "\n" + (
 "What are the effective treatments or solutions for addressing sudden patchy hair loss,"
 "commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
)
response(user_input_3)

Llama.generate: prefix-match hit


"\n\nRelevant parts of context:\n1. Sudden patchy hair loss\n2. Localized bald spots on the scalp\n3. Effective treatments or solutions\n4. Possible causes\n\nAnswer:\nThe sudden appearance of localized bald spots on the scalp, also known as alopecia areata, can be distressing. The exact cause is unknown but it's believed to involve an immune system reaction that attacks hair follicles. Some possible triggers include stress, genetics, and viral infections.\n\nEffective treatments or solutions for addressing sudden patchy hair loss include:\n1. Topical immunotherapy: This involves applying a medication directly to the affected area to stimulate hair regrowth.\n2. Corticosteroids: These drugs can be applied topically, injected into the scalp, or taken orally to reduce inflammation and suppress the immune system response.\n3. Minoxidil: This medication is available over-the-counter and can help slow down hair loss and promote regrowth.\n4. Hair transplant: In severe cases where there is e

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
# Defining the user input for "What treatments are recommended for a person who has sustained a physical injury
# to brain tissue, resulting in temporary or permanent impairment of brain function?" and getting the response
user_input_4 = system_prompt + "\n" + (
    "What treatments are recommended for a person who has sustained a physical injury"
    "to brain tissue, resulting in temporary or permanent impairment of brain function?"
)
response(user_input_4)

Llama.generate: prefix-match hit


"\n\nBased on current medical knowledge, the treatment for a brain injury depends on the severity and location of the injury. Here are some common treatments:\n\n1. Emergency care: For severe injuries, immediate medical attention is necessary to prevent further damage. This may include surgery to remove hematomas or decompressing skull fractures.\n2. Medications: Depending on the symptoms, doctors may prescribe medications to manage seizures, control pain, reduce swelling, or improve cognitive function.\n3. Rehabilitation: Physical therapy, occupational therapy, speech therapy, and other forms of rehabilitation can help individuals regain lost skills and improve overall functioning.\n4. Assistive devices: Devices such as wheelchairs, walkers, braces, or communication aids may be recommended to help individuals with mobility or communication challenges.\n5. Lifestyle modifications: Making lifestyle changes, such as getting enough rest, eating a healthy diet, and avoiding alcohol and dru

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
# Defining the user input for "What are the necessary precautions and treatment steps for a person who has
# fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
# and getting the response
user_input_5 = system_prompt + "\n" + (
    "What are the necessary precautions and treatment steps for a person who has fractured their leg during"
    "a hiking trip, and what should be considered for their care and recovery?"
)
response(user_input_5)

Llama.generate: prefix-match hit


'\n\nTo provide an accurate response, I need to clarify that we\'re discussing a diagonally broken long bone in the lower limb, commonly referred to as a "fractured leg." Here are the necessary precautions and treatment steps:\n\n1. **Immediate Care:** If the person is in severe pain or unable to move, immobilize the leg using a splint or sling. Do not attempt to realign the bone yourself. Seek medical help as soon as possible.\n\n2. **Transportation:** Transport the person to the hospital using a level surface and avoid jarring movements that could worsen the injury. If necessary, call for emergency medical services.\n\n3. **Hospital Treatment:** The doctor may perform an X-ray to confirm the fracture\'s location and severity. Depending on the type and location of the fracture, they might apply a cast or use traction to realign the bone. In some cases, surgery may be required to insert metal plates or screws to hold the bone in place.\n\n4. **Medication:** The doctor may prescribe pai

**Observation**


*   The answers become more specific
*   List item


## Data Preparation for RAG

### Importing Necessary Libraries

In [ ]:
# Importing required libraries

#Libraries for processing dataframes,text
import json,os
import tiktoken
import pandas as pd

#Libraries for Loading Data, Chunking, Embedding, and Vector Databases
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.document_loaders import PyMuPDFLoader
from langchain_community.embeddings.sentence_transformer import SentenceTransformerEmbeddings
from langchain_community.vectorstores import Chroma

from huggingface_hub import hf_hub_download
from llama_cpp import Llama

### Loading the Data

In [ ]:
# Mounting  Google Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Setting the Google Drive path for medical journal
medical_journal_pdf_path = "/content/drive/MyDrive/Learning/UT Austin/05 - Natural Language Processing with Generative AI/Project/medical_diagnosis_manual.pdf"

In [ ]:
# Loading the medical journal
pdf_loader = PyMuPDFLoader(medical_journal_pdf_path)
medical_jorunal = pdf_loader.load()

### Data Overview

#### Checking the first 5 pages

In [ ]:
# Checking first 5 pages
for i in range(5):
    print(f"Page Number: {i+1}")
    print()
    print(medical_jorunal[i].page_content)
    print()

Page Number: 1

tzmjoseph@gmail.com
G8OFHRLDZN
This file is meant for personal use by tzmjoseph@gmail.com only.
Sharing or publishing the contents in part or full is liable for legal action.

Page Number: 2

tzmjoseph@gmail.com
G8OFHRLDZN
This file is meant for personal use by tzmjoseph@gmail.com only.
Sharing or publishing the contents in part or full is liable for legal action.

Page Number: 3

Table of Contents
1
Front    ................................................................................................................................................................................................................
1
Cover    .......................................................................................................................................................................................................
2
Front Matter    .....................................................................................................................................

**Observation**


*   The page contents are read correctly.

#### Checking the number of pages

In [ ]:
# Print the number of pages
print(f"Number of pages: {len(medical_jorunal)}")

Number of pages: 4114


**Observation**


*   It has correctly identified the number of pages (4114) in the document.



### Data Chunking

In [ ]:
# Create text splitter to chunk the data with chink size of 512 and overlap around 5% of chunk size
text_splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
    encoding_name='cl100k_base',
    chunk_size=512,
    chunk_overlap=25
)

In [ ]:
# Create document chunks using above text splitter
document_chunks = pdf_loader.load_and_split(text_splitter)

In [ ]:
# Printing number of document chunks
print(f"Number of document chunks: {len(document_chunks)}")

Number of document chunks: 8531


In [ ]:
# Checking first chunk's page contents
print(document_chunks[0].page_content)

tzmjoseph@gmail.com
G8OFHRLDZN
This file is meant for personal use by tzmjoseph@gmail.com only.
Sharing or publishing the contents in part or full is liable for legal action.


In [ ]:
# Checking 98th chunk's page contents
print(document_chunks[98].page_content)

Diagnosis is based on results of medical and diet histories, physical examination, body composition
analysis (see p. 58), and selected laboratory tests.
History: History should include questions about dietary intake (see
Fig. 2-1), recent changes in weight, and risk factors for undernutrition, including drug and alcohol use.
Unintentional loss of ≥ 10% of usual body weight during a 3-mo period indicates a high probability of
undernutrition. Social history should include questions about whether money is available for food and
whether the patient can shop and cook.
Review of systems should focus on symptoms of nutritional deficiencies (see Table 2-1). For example,
impaired night vision may indicate vitamin A deficiency.
Physical examination: Physical examination should include measurement of height and weight,
inspection of body fat distribution, and anthropometric measurements of lean body mass. Body mass
index (BMI = weight(kg)/height(m)2) adjusts weight for height (see
Table 6-2 on p.

In [ ]:
# Checking 99th chunk's page contents
print(document_chunks[99].page_content)

Physical examination should focus on signs of specific nutritional deficiencies. Signs of PEU (eg, edema,
muscle wasting, skin changes) should be sought. Examination should also focus on signs of conditions
that could predispose to nutritional deficiencies, such as dental problems. Mental status should be
assessed, because depression and cognitive impairment can lead to weight loss.
The widely used Subjective Global Assessment (SGA) uses information from the patient history (eg,
weight loss, change in intake, GI symptoms), physical examination findings (eg, loss of muscle and
subcutaneous fat, edema, ascites), and the clinician's judgment of the patient's nutritional status. The
Mini Nutritional Assessment (MNA) has been validated and is widely used, especially for elderly patients
(see Fig. 2-1). The Simplified Nutrition Assessment Questionnaire (SNAQ), a simple, validated method of
predicting future weight loss, may be used (see
Fig. 2-2).
Testing: The extent of laboratory testing ne

In [ ]:
# Checking 100th chunk's page contents
print(document_chunks[100].page_content)

mortality better than measurement of the other proteins. However, the correlation of albumin with
morbidity and mortality may be related to nonnutritional as well as nutritional factors. Inflammation
produces cytokines that cause albumin and other nutritional protein markers to extravasate, decreasing
serum levels. Because prealbumin, transferrin, and retinol-binding protein decrease more rapidly during
starvation than does albumin, their measurements are sometimes used to diagnose or assess the severity
of acute starvation. However, whether they are more sensitive or specific than albumin is unclear.
Total lymphocyte count, which often decreases as undernutrition progresses, may be determined.
Undernutrition causes a marked decline in CD4+ T lymphocytes, so this count may not be useful in
patients who have AIDS.
Skin tests using antigens can detect impaired cell-mediated immunity in PEU and in some other disorders
of undernutrition (see p.
1098).
Other laboratory tests, such as measur

In [ ]:
# Checking 101th chunk's page contents
print(document_chunks[101].page_content)

90%; moderate, 75 to 85%; severe, <75%).
PEU may be primary or secondary. Primary PEU is caused by inadequate nutrient intake. Secondary PEU
results from disorders or drugs that interfere with nutrient use.
Primary PEU: Worldwide, primary PEU occurs mostly in children and the elderly who lack access to
nutrients, although a common cause in the elderly is depression. PEU can also result from fasting or
anorexia nervosa. Child or elder abuse may be a cause.
In children, chronic primary PEU has 2 common forms: marasmus and kwashiorkor. The form depends on
the balance of nonprotein and protein sources of energy. Starvation is an acute severe form of primary
PEU.
Marasmus (also called the dry form of PEU) causes weight loss and depletion of fat and muscle. In
developing countries, marasmus is the most common form of PEU in children.
Kwashiorkor (also called the wet, swollen, or edematous form) is associated with premature abandonment
The Merck Manual of Diagnosis & Therapy, 19th Edition
Cha

In [ ]:
# Checking third last chunk's page contents
print(document_chunks[-3].page_content)

X
Xanthelasma 579
Xanthine oxidase deficiency 3026
Xanthines 2064
Xanthochromia 1595
Xanthogranulomatous pyelonephritis 2379
Xanthomas 634, 645, 895
Xanthomatosis, cerebrotendinous 894
X chromosome 2599, 3373 (see also Genetic)
abnormalities of 3002-3005, 3376-3377, 3377
inactivation of 3002, 3378
47,XXX syndrome 3005
48,XXXX syndrome 3005
49,XXXXX syndrome 3005
47,XYY syndrome 3005
Xenograft 1126 (see also Transplantation)
Xeroderma (xerosis) 661
Xerophthalmia 594
vitamin A deficiency and 34
Xerosis 636
Xerostomia 501, 513-516
radiation-induced 488, 505
in Sjogren's syndrome 304
X-linked adrenoleukodystrophy 3024
X-linked agammaglobulinemia 1092, 1097, 1107-1108
X-linked lymphoproliferative syndrome 1092, 1097, 1108
X-rays 3252 (see also Radiation; Radiation therapy; Radiography)
Xylene poisoning 3348
Xylitol 518
Xylose absorption test 155, 3494, 3500
Xylose breath test 156, 157
The Merck Manual of Diagnosis & Therapy, 19th Edition
X
4102
tzmjoseph@gmail.com
G8OFHRLDZN
This file is me

In [ ]:
# Checking second last chunk's page contents
print(document_chunks[-2].page_content)

Y
Yaws 1266-1267
forest 1379
Y chromosome 3373 (see also Genetic)
abnormalities of 3005
Yeast infection (see also Fungal infection)
vaginal 2542, 2544, 2545
Yellow fever 1400, 1429, 1437
hepatic inflammation in 248
vaccine against 1172, 1437, 3441
Yellow nail syndrome 732, 1995
pleural effusion in 1997
Yellow skin (see Jaundice)
Yersinia infection 1167, 1256-1257
Y. enterocolitica infection 147
Y. pestis infection 1924
Yew poisoning 3338
Yips 1762
Yo, antibodies to 1056
Yolk sac tumor 2476
The Merck Manual of Diagnosis & Therapy, 19th Edition
Y
4103
tzmjoseph@gmail.com
G8OFHRLDZN
This file is meant for personal use by tzmjoseph@gmail.com only.
Sharing or publishing the contents in part or full is liable for legal action.


In [ ]:
# Checking last chunk's page contents
print(document_chunks[-1].page_content)

Z
Zafirlukast 1879
Zalcitabine 1451
in children 2854
Zaleplon 1709
Zanamivir 1407
in influenza 1407, 1929
ZAP-70 (zeta-associated protein 70) deficiency 1092, 1108
Zavanelli maneuver 2680
Zellweger syndrome 2383, 3023
Zenker's diverticulum 125
Zidovudine 1451, 1453
in children 2854
Zileuton 1881
in asthma 1880
Zinc 49, 55, 3431-3432
in common cold 1405
deficiency of 11, 49, 55
in dermatophytoses 705
poisoning with 3328, 3353
recommended dietary allowances for 50
reference values for 3499
toxicity of 49, 55
copper deficiency and 49
in Wilson's disease 52
Zinc oxide 2233
gelatin formulation of 646, 672
Zinc pyrithione 647
Zinc shakes 55
Zipper injury 3239, 3240
Ziprasidone
in agitation 1492
in bipolar disorder 3059
poisoning with 3347
in schizophrenia 1566
Zoledronate 359, 361, 848
Zollinger-Ellison syndrome 95, 199, 200-201, 910
mastocytosis vs 1125
Menetrier's disease vs 132
peptic ulcer disease vs 134
Zolmitriptan 1721
Zolpidem 1709, 3103
Zonisamide 1701
Zoonotic diseases, cutaneous 7

**Observation**


*   The model has created 8531 chunks
*   There are some overlaps



### Embedding

In [ ]:
# Creating embedding
embedding_model = SentenceTransformerEmbeddings(model_name="thenlper/gte-large")

/tmp/ipykernel_1846/2821821798.py:2: LangChainDeprecationWarning: The class `HuggingFaceEmbeddings` was deprecated in LangChain 0.2.2 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-huggingface package and should be used instead. To use it run `pip install -U :class:`~langchain-huggingface` and import as `from :class:`~langchain_huggingface import HuggingFaceEmbeddings``.
  embedding_model = SentenceTransformerEmbeddings(model_name="thenlper/gte-large")


modules.json:   0%|          | 0.00/385 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/57.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/619 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/670M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/342 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/191 [00:00<?, ?B/s]

In [ ]:
# Creating embeddings for 1st and 2nd chunk
embedding_1 = embedding_model.embed_query(document_chunks[0].page_content)
embedding_2 = embedding_model.embed_query(document_chunks[1].page_content)

In [ ]:
# Printing dimensions of embedding
print(f"Dimension of embedding: {len(embedding_1)}")

# Checking if dimensions of embedding are same for both chunks
dimensions_same = len(embedding_1) == len(embedding_2)
print(f"Are dimeshions same?: {dimensions_same}")

Dimension of embedding: 1024
Are dimeshions same?: True


**Observation**


*   The dimensions of the created embedding are 1024 across various chunks



### Vector Database

In [ ]:
# Creating directory for embedding
out_dir = 'medical_journal_db'

if not os.path.exists(out_dir):
    os.makedirs(out_dir)

In [ ]:
# Creating vector store
vectorstore = Chroma.from_documents(
    document_chunks,
    embedding_model,
    persist_directory=out_dir
)

In [ ]:
vectorstore = Chroma(persist_directory=out_dir,embedding_function=embedding_model)

/tmp/ipykernel_1846/2756559696.py:1: LangChainDeprecationWarning: The class `Chroma` was deprecated in LangChain 0.2.9 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-chroma package and should be used instead. To use it run `pip install -U :class:`~langchain-chroma` and import as `from :class:`~langchain_chroma import Chroma``.
  vectorstore = Chroma(persist_directory=out_dir,embedding_function=embedding_model)


In [ ]:
vectorstore.embeddings

HuggingFaceEmbeddings(client=SentenceTransformer(
  (0): Transformer({'max_seq_length': 512, 'do_lower_case': False, 'architecture': 'BertModel'})
  (1): Pooling({'word_embedding_dimension': 1024, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
  (2): Normalize()
), model_name='thenlper/gte-large', cache_folder=None, model_kwargs={}, encode_kwargs={}, multi_process=False, show_progress=False)

In [ ]:
vectorstore.similarity_search("Diabetes Type 2 Symptoms",k=3)

[Document(metadata={'author': '', 'file_path': '/content/drive/MyDrive/Learning/UT Austin/05 - Natural Language Processing with Generative AI/Project/medical_diagnosis_manual.pdf', 'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)', 'creator': 'Atop CHM to PDF Converter', 'modDate': 'D:20260424043741Z', 'subject': '', 'page': 1012, 'creationDate': 'D:20120615054440Z', 'source': '/content/drive/MyDrive/Learning/UT Austin/05 - Natural Language Processing with Generative AI/Project/medical_diagnosis_manual.pdf', 'creationdate': '2012-06-15T05:44:40+00:00', 'total_pages': 4114, 'trapped': '', 'keywords': '', 'moddate': '2026-04-24T04:37:41+00:00', 'title': 'The Merck Manual of Diagnosis & Therapy, 19th Edition', 'format': 'PDF 1.7'}, page_content='The most common symptoms of DM are those of hyperglycemia: an osmotic diuresis caused by\nglycosuria leading to urinary frequency, polyuria, and polydipsia that may progress to orthostatic\nhypotension and dehydration. Severe dehydration 

### Retriever

In [ ]:
# Creating similarity retriever
retriever = vectorstore.as_retriever(
    search_type='similarity',
    search_kwargs={'k': 2}
)

In [ ]:
rel_docs = retriever.get_relevant_documents("What are the symptoms of Type 2 Diabetes?")
rel_docs

/tmp/ipykernel_1846/1567604316.py:1: LangChainDeprecationWarning: The method `BaseRetriever.get_relevant_documents` was deprecated in langchain-core 0.1.46 and will be removed in 1.0. Use :meth:`~invoke` instead.
  rel_docs = retriever.get_relevant_documents("What are the symptoms of Type 2 Diabetes?")


[Document(metadata={'keywords': '', 'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)', 'title': 'The Merck Manual of Diagnosis & Therapy, 19th Edition', 'total_pages': 4114, 'moddate': '2026-04-24T04:37:41+00:00', 'source': '/content/drive/MyDrive/Learning/UT Austin/05 - Natural Language Processing with Generative AI/Project/medical_diagnosis_manual.pdf', 'format': 'PDF 1.7', 'creationdate': '2012-06-15T05:44:40+00:00', 'modDate': 'D:20260424043741Z', 'page': 1012, 'creationDate': 'D:20120615054440Z', 'trapped': '', 'file_path': '/content/drive/MyDrive/Learning/UT Austin/05 - Natural Language Processing with Generative AI/Project/medical_diagnosis_manual.pdf', 'creator': 'Atop CHM to PDF Converter', 'author': '', 'subject': ''}, page_content='The most common symptoms of DM are those of hyperglycemia: an osmotic diuresis caused by\nglycosuria leading to urinary frequency, polyuria, and polydipsia that may progress to orthostatic\nhypotension and dehydration. Severe dehydration 

### System and User Prompt Template

In [ ]:
# Creating system message
qna_system_message = """
You are an assistant whose work is to review the report and provide the appropriate answers from the context.
User input will have the context required by you to answer user questions.
This context will begin with the token: ###Context.
The context contains references to specific portions of a document relevant to the user query.

User questions will begin with the token: ###Question.

Please answer only using the context provided in the input. Do not mention anything about the context in your final answer.

If the answer is not found in the context, respond "I don't know".
"""

In [ ]:
qna_user_message_template = """
###Context
Here are some documents that are relevant to the question mentioned below.
{context}

###Question
{question}
"""

### Response Function

In [ ]:
def generate_rag_response(user_input,k=3,max_tokens=3500,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=k)
    context_list = [d.page_content for d in relevant_document_chunks]

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = qna_system_message + '\n' + user_message

    # Generate the response
    try:
        response = llm(
                  prompt=prompt,
                  max_tokens=max_tokens,
                  temperature=temperature,
                  top_p=top_p,
                  top_k=top_k
                  )

        # Extract and print the model's response
        response = response['choices'][0]['text'].strip()
    except Exception as e:
        response = f'Sorry, I encountered the following error: \n {e}'

    return response

## Question Answering using RAG

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
# Defining user input for "What is the protocol for managing sepsis in a critical care unit?" and getting the response
user_input_1 = "What is the protocol for managing sepsis in a critical care unit?"
print(generate_rag_response(user_input_1))

Llama.generate: prefix-match hit


Based on the context provided, the protocol for managing sepsis in a critical care unit includes:
1. Parenteral antibiotics should be given after specimens of blood, body fluids, and wound sites have been taken for Gram stain and culture.
2. Very prompt empiric therapy, started immediately after suspecting sepsis, is essential and may be lifesaving.
3. Antibiotic selection requires an educated guess based on the suspected source, clinical setting, knowledge or suspicion of causative organisms and of sensitivity patterns common to that specific inpatient unit, and previous culture results.
4. One regimen for septic shock of unknown cause is gentamicin or tobramycin 5.1 mg/kg IV once/day plus a 3rd-generation cephalosporin (cefotaxime 2 g q 6 to 8 h or ceftriaxone 2 g once/day) or ceftazidime 2 g IV q 8 h, or ceftazidime plus a fluoroquinolone (eg, ciprofloxacin).
5. Vancomycin must be added if resistant staphylococci or enterococci are suspected. If there is an abdominal source, a drug 

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
# Defining the user input for "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what
# surgical procedure should be followed to treat it?" and getting the response
user_input_2 = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
print(generate_rag_response(user_input_2))

Llama.generate: prefix-match hit


The common symptoms for appendicitis include abdominal pain, anorexia (loss of appetite), and abdominal tenderness. Appendicitis cannot be cured via medicine alone as it requires surgical removal of the appendix. The preferred surgical procedures are open or laparoscopic appendectomy.


### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
# Defining the user input for "What are the effective treatments or solutions for addressing sudden patchy hair loss,
# commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?" and getting the response
user_input_3 = (
 "What are the effective treatments or solutions for addressing sudden patchy hair loss,"
 "commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
)
print(generate_rag_response(user_input_3))

Llama.generate: prefix-match hit


Based on the context provided, the condition being described is alopecia areata. The treatment options for this condition include topical, intralesional, or systemic corticosteroids, topical minoxidil, topical anthralin, topical immunotherapy (diphencyprone or squaric acid dibutylester), or psoralen plus ultraviolet A (PUVA). The cause of alopecia areata is thought to be an autoimmune disorder affecting genetically susceptible people exposed to unclear environmental triggers.


### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
# Defining the user input for "What treatments are recommended for a person who has sustained a physical injury
# to brain tissue, resulting in temporary or permanent impairment of brain function?" and getting the response
user_input_4 = (
    "What treatments are recommended for a person who has sustained a physical injury"
    "to brain tissue, resulting in temporary or permanent impairment of brain function?"
)
print(generate_rag_response(user_input_4))

Llama.generate: prefix-match hit


Based on the context provided, the recommended treatments for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function include:
1. Ensuring a reliable airway and maintaining adequate ventilation, oxygenation, and blood pressure.
2. Surgery to place monitors to track and treat intracranial pressure, decompress the brain if intracranial pressure is increased, or remove intracranial hematomas.
3. Maintaining adequate brain perfusion and oxygenation and preventing complications of altered sensorium in the first few days after the injury.
4. Rehabilitation for many patients subsequently.
5. Preventing systemic complications due to immobilization, such as pneumonia, UTI, and thromboembolic disease.
6. Providing good nutrition.
7. Preventing pressure ulcers.


### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
# Defining the user input for "What are the necessary precautions and treatment steps for a person who has
# fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
# and getting the response
user_input_5 = (
    "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip,"
    "and what should be considered for their care and recovery?"
)
print(generate_rag_response(user_input_5))

Llama.generate: prefix-match hit


Based on the context provided, here's the answer:

The person with a fractured leg should follow these steps:
1. Elevate the injured limb above heart level for the first 2 days to minimize swelling.
2. Apply warmth (using a heating pad) for 15 to 20 minutes after the first 48 hours to relieve pain and speed healing.
3. Immobilize the joints proximal and distal to the injury using elastic bandages, a cast, or a splint.
   - If a cast is used, keep it dry, never put anything inside it, inspect its edges and skin daily, apply lotion to any red or sore areas, pad rough edges with soft material, and seek medical care if an odor emanates from the cast or a fever develops.
4. Apply ice to the injury for pain relief and swelling reduction.
5. If the fracture requires prolonged immobilization (more than 3 to 4 weeks), prevent complications such as stiffness, contractures, and muscle atrophy by resuming active motion within the first few days or weeks (early mobilization).
6. For severe swelling

### Fine-tuning

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
# Defining user input for "What is the protocol for managing sepsis in a critical care unit?" and getting the response
user_input_1 = "What is the protocol for managing sepsis in a critical care unit?"
print(generate_rag_response(user_input_1, temperature=0.5))

Llama.generate: prefix-match hit


Based on the context provided, the protocol for managing sepsis in a critical care unit includes the following steps:
1. Administer parenteral antibiotics as soon as possible after suspecting sepsis based on suspected source, clinical setting, and previous culture results. Empiric therapy should be started immediately.
2. Monitor physiologic parameters frequently including systemic pressure, CVP or PAOP, pulse oximetry, arterial blood gases, blood glucose, lactate, electrolyte levels, renal function, and possibly sublingual PCO2. Urine output should also be measured.
3. Provide adequate fluid resuscitation with 0.9% saline until CVP reaches 8 mm Hg or PAOP reaches 12 to 15 mm Hg. Oliguria with hypotension is not a contraindication to vigorous fluid resuscitation.
4. If the patient remains hypotensive after target levels have been reached, dopamine may be given to increase mean BP to at least 60 mm Hg. If necessary, another vasopressor like norepinephrine can be added but care should be

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
# Defining the user input for "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what
# surgical procedure should be followed to treat it?" and getting the response
user_input_2 = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
print(generate_rag_response(user_input_2, temperature=0.5))

Llama.generate: prefix-match hit


Appendicitis is a condition characterized by inflammation of the vermiform appendix. The typical symptoms include abdominal pain, anorexia (loss of appetite), and abdominal tenderness. However, these symptoms can also be present in other conditions, so a definitive diagnosis requires clinical evaluation and imaging studies such as CT or ultrasound.

No, appendicitis cannot be cured via medicine alone. Surgical removal of the appendix, known as appendectomy, is the standard treatment for appendicitis to prevent complications like rupture, peritonitis, and sepsis. Antibiotics are usually administered before surgery to reduce the bacterial load in the abdominal cavity and improve post-operative recovery. In cases where surgery is impossible, antibiotics can help improve survival rates but do not cure the condition.


### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
# Defining the user input for "What are the effective treatments or solutions for addressing sudden patchy hair loss,
# commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?" and getting the response
user_input_3 = (
 "What are the effective treatments or solutions for addressing sudden patchy hair loss,"
 "commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
)
print(generate_rag_response(user_input_3, temperature=0.5))

Llama.generate: prefix-match hit


Based on the context provided, the condition being described is alopecia areata. The treatment options for this condition include topical, intralesional, or systemic corticosteroids, topical minoxidil, topical anthralin, topical immunotherapy (diphencyprone or squaric acid dibutylester), or psoralen plus ultraviolet A (PUVA). The cause of alopecia areata is thought to be an autoimmune disorder affecting genetically susceptible people exposed to unclear environmental triggers.


### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
# Defining the user input for "What treatments are recommended for a person who has sustained a physical injury
# to brain tissue, resulting in temporary or permanent impairment of brain function?" and getting the response
user_input_4 = (
    "What treatments are recommended for a person who has sustained a physical injury"
    "to brain tissue, resulting in temporary or permanent impairment of brain function?"
)
print(generate_rag_response(user_input_4, temperature=0.5))

Llama.generate: prefix-match hit


I don't know without specific information about whether the injury is open or closed and the severity of the impairment. However, based on the context provided, it appears that supportive care is recommended for all patients with traumatic brain injuries. This includes preventing systemic complications due to immobilization, providing good nutrition, and preventing pressure ulcers. Additionally, rehabilitation specialists are important for maximal functional recovery, which includes prevention of secondary disabilities, such as pressure ulcers, joint contractures, and pneumonia. For patients with severe cognitive dysfunction, extensive cognitive therapy may be required.


### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
# Defining the user input for "What are the necessary precautions and treatment steps for a person who has
# fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
# and getting the response
user_input_5 = (
    "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip,"
    "and what should be considered for their care and recovery?"
)
print(generate_rag_response(user_input_5, temperature=0.5))

Llama.generate: prefix-match hit


The context mentions that fractures involve cracks in bones and symptoms include pain, swelling, ecchymosis, crepitation, deformity, and abnormal motion. The treatment includes analgesics, immobilization, and sometimes surgery. Immobilization can be done using a cast or a splint. A cast is typically used for fractures that require weeks of immobilization. Patients with casts should keep them dry, not put anything inside, inspect the edges daily, and seek medical care if there's an odor or fever. Splints can be used for stable injuries that require less immobilization and do not contribute to compartment syndrome. Prolonged immobilization can cause complications such as stiffness, contractures, and muscle atrophy. Some rapidly healing injuries benefit from early mobilization. Prehospital care for a fractured limb includes rewarming the affected area in tolerably warm water and providing analgesics if available. Acute care involves stabilizing core temperature and rapidly rewarming the l

## Output Evaluation

Let us now use the LLM-as-a-judge method to check the quality of the RAG system on two parameters - retrieval and generation. We illustrate this evaluation based on the answeres generated to the question from the previous section.

- We are using the same Mistral model for evaluation, so basically here the llm is rating itself on how well he has performed in the task.

In [ ]:
groundedness_rater_system_message = """
Rate groundedness from 1-5.

Metric:
Answer must be derived only from context.

Return ONLY:
Score: <1-5>
Reason: <max 2 sentences>
"""

In [ ]:
relevance_rater_system_message = """
Rate relevance from 1-5.

Metric:
How well answer addresses question using context.

Return ONLY:
Score: <1-5>
Reason: <max 2 sentences>
"""

In [ ]:
user_message_template = """
###Question
{question}

###Context
{context}

###Answer
{answer}
"""

In [ ]:
def generate_ground_relevance_response(user_input,k=3,max_tokens=3500,temperature=0,top_p=0.95,top_k=5):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=3)
    context_list = [d.page_content for d in relevant_document_chunks]
    context_for_query = ". ".join(context_list)

    # Combine user_prompt and system_message to create the prompt
    prompt = f"""[INST]{qna_system_message}\n
                {'user'}: {qna_user_message_template.format(context=context_for_query, question=user_input)}
                [/INST]"""

    response = llm(
            prompt=prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    answer =  response["choices"][0]["text"]


    # Combine user_prompt and system_message to create the prompt
    groundedness_prompt = f"""[INST]{groundedness_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    # Combine user_prompt and system_message to create the prompt
    relevance_prompt = f"""[INST]{relevance_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    response_1 = llm(
            prompt=groundedness_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    response_2 = llm(
            prompt=relevance_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    return response_1['choices'][0]['text'],response_2['choices'][0]['text']

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
# Defining user input for "What is the protocol for managing sepsis in a critical care unit?" and getting the response
user_input_1 = "What is the protocol for managing sepsis in a critical care unit?"
ground,rel = generate_ground_relevance_response(user_input_1, max_tokens=3500)

print(ground, end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Score: 5
Reason: The answer is derived directly from the context, specifically the sections discussing the management of sepsis in an ICU and the monitoring and treatment strategies for septic shock patients.

 Score: 5
Reason: The answer directly addresses the question by summarizing the key steps in managing sepsis in a critical care unit as outlined in the context, including monitoring physiologic parameters, fluid resuscitation, administration of parenteral antibiotics, and surgical intervention when necessary.


### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
# Defining the user input for "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what
# surgical procedure should be followed to treat it?" and getting the response
user_input_2 = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
ground,rel = generate_ground_relevance_response(user_input_2, max_tokens=3500)

print(ground, end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Score: 5
Reason: The context directly answers the question about the common symptoms for appendicitis and states that the standard treatment is surgical removal (appendectomy).

 Score: 5
Reason: The answer directly addresses the user's question by listing the common symptoms for appendicitis and stating that it cannot be cured with medicine, but rather requires surgical removal (appendectomy). The context provided in the text fully supports this information.


### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
# Defining the user input for "What are the effective treatments or solutions for addressing sudden patchy hair loss,
# commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?" and getting the response
user_input_3 = (
 "What are the effective treatments or solutions for addressing sudden patchy hair loss,"
 "commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
)
ground,rel = generate_ground_relevance_response(user_input_3, max_tokens=3500)

print(ground, end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Score: 5
Reason: The answer is derived directly from the context provided, which discusses various treatments for alopecia areata and mentions that it is an autoimmune disorder.

 Score: 5
Reason: The answer directly addresses the question by providing information about effective treatments for sudden patchy hair loss, which is a key point from the context mentioned as alopecia areata. The answer also explains that this type of hair loss is believed to be an autoimmune disorder and provides examples of possible treatments.


### Query 4: What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
# Defining the user input for "What treatments are recommended for a person who has sustained a physical injury
# to brain tissue, resulting in temporary or permanent impairment of brain function?" and getting the response
user_input_4 = (
    "What treatments are recommended for a person who has sustained a physical injury"
    "to brain tissue, resulting in temporary or permanent impairment of brain function?"
)
ground,rel = generate_ground_relevance_response(user_input_4, max_tokens=3500)

print(ground, end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Score: 1
Reason: The context does not mention any specific medical treatments or medications for a person with a brain injury. It focuses on the role of rehabilitation specialists, preventive measures, and therapy types.

 Score: 1
Reason: The context does not directly address the question about specific treatments for a person with a brain injury. It focuses on the role of rehabilitation specialists, preventive measures, and different types of injuries.


### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
# Defining the user input for "What are the necessary precautions and treatment steps for a person who has
# fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
# and getting the response
user_input_5 = (
    "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip,"
    "and what should be considered for their care and recovery?"
)
ground,rel = generate_ground_relevance_response(user_input_5, max_tokens=3500)

print(ground, end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


 Score: 5
Reason: The answer is completely derived from the context provided, which includes information on necessary precautions, treatment steps, and considerations for care and recovery for a person with a fractured leg.

 Score: 5
Reason: The answer directly addresses the question by providing necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, using information from the context provided.


## Actionable Insights and Business Recommendations


*   Given this is related to medical diagnosis it is better to use lower temperature to control the randomness in the answers.
*   The prompt for groundedness and relevance has to be succinct to allow the system to provide a rating. A large prompt results in the output being truncated.
*   Given the source is a single document a lower k value (3) would suffice.



<font size=6 color='blue'>Power Ahead</font>
___

In [ ]:
import json

input_path = "/content/drive/MyDrive/Learning/UT Austin/05 - Natural Language Processing with Generative AI/Project/ThomasJoseph_Full_Code_NLP_RAG_Project_Notebook.ipynb"
clean_path = "/content/drive/MyDrive/Learning/UT Austin/05 - Natural Language Processing with Generative AI/Project/ThomasJoseph_Full_Code_NLP_RAG_Project_Notebook_no_widget_metadata.ipynb"

with open(input_path, "r", encoding="utf-8") as f:
    nb = json.load(f)

# Remve notebook level widget metadata that is causing KeyError: 'state'
if "metadata" in nb and "widgets" in nb["metadata"]:
    del nb["metadata"]["widgets"]

with open(clean_path, "w", encoding="utf-8") as f:
    json.dump(nb, f, ensure_ascii=False, indent=1)

print("Saved cleaned notebook:", clean_path)
!jupyter nbconvert --to html "/content/drive/MyDrive/Learning/UT Austin/05 - Natural Language Processing with Generative AI/Project/ThomasJoseph_Full_Code_NLP_RAG_Project_Notebook_no_widget_metadata.ipynb"

Saved cleaned notebook: /content/drive/MyDrive/Learning/UT Austin/05 - Natural Language Processing with Generative AI/Project/ThomasJoseph_Full_Code_NLP_RAG_Project_Notebook_no_widget_metadata.ipynb
[NbConvertApp] Converting notebook /content/drive/MyDrive/Learning/UT Austin/05 - Natural Language Processing with Generative AI/Project/ThomasJoseph_Full_Code_NLP_RAG_Project_Notebook_no_widget_metadata.ipynb to html
[NbConvertApp] Writing 548123 bytes to /content/drive/MyDrive/Learning/UT Austin/05 - Natural Language Processing with Generative AI/Project/ThomasJoseph_Full_Code_NLP_RAG_Project_Notebook_no_widget_metadata.html
